txtl/mixture.py

```diff
@@ -0,0 +1,163 @@
+    concentrations      Concentration of each component (list of floats)
+    default_mechanisms  Default mechanisms for this mixture (dict)
+    custom_mechanisms   User-specified mechanisms for this mixture (dict)
+    parameters          Global parameters for the mixture (dict)
+
+    The mechanisms and parameters dictionaries are established by the
+    create_extract() and create_buffer() functions, using the
+    properties of the Extract and Buffer components.
+
+    Operations on mixtures
+    ----------------------
+    # create a new mixture
+    mix = create_mixture("name")
+
+    # add a DNA assembly to a mixture
+    add_dna(mix, dna_assy, conc)        
+    mix += conc * dna_assy
+
+    # Combine mixtures at scaled concentrations
+    mix = combine_mixtures([mix1, mix2, ...], [conc1, conc2, ...]
+    mix = conc1 * mix1 + conc2 * mix2 + conc3 * mix3
+
+    # Add a component to a mixture
+    mix = mix + conc * component
+    mix += conc * component
+
+    # Generate an SMBL file for the model represented in a mixture
+    write_sbml(mix, filename)
+    mix.write_sbml(filename)
+
+    """
+    def __init__(self, name=None, mechanisms={}, config_file=None):
+        "Create a new mixture"
+        
+        # Create an SBML document container and model
+        document, model, compartment = create_sbml_model()
+
+        # Initialize instance variables
+        self.name = name;               # Save the name of the mixture
+        self._SBMLdoc = document        # SBML document container
+        self.model = model              # SBML model object
+        self.compartment = compartment  # SBML compartment
+        self.components = []            # components contained in mixture
+        self.concentrations = []        # concentrations of each component
+
+        # Override the default mechanisms with anything we were passed
+        # Note: These are overrwriten by create_extract()
+        self.default_mechanisms = {
+            # 'mechanism' : MechanismConstructor()
+        }
+        self.custom_mechanisms = mechanisms
+
+        # Read the configuration parameters
+        self.parameters = {}
+        if (config_file != None):
+            self.parameters.update(load_config(config_file))
+
+    def write_sbml(self, filename):
+        "Generate an SBML file for the current mixture (model)"
+        
+        # Update all species in the mixture to make sure everything exists
+        for i in range(len(self.components)):
+            concentration = self.concentrations[i]
+            component = self.components[i]
+
+            # Create all of the species for this component
+            component.update_species(self, concentration)
+
+        # Now go through and add all of the reactions that are required
+        for component in self.components:
+            component.update_reactions(self)
+
+        # Write the model to a file
+        libsbml.writeSBMLToFile(self._SBMLdoc, filename)
```

Comment:
We could possibly put all models generated in SBML in a common subfolder called models in our directory?